### ISSUE TITLE
Getting Exception "Invalid pulses per quarter note value." for every midi file I try to load.

### ISSUE THREAD

**Issue by MaxxDelusional** [tebjan/Sanford.Multimedia.Midi#2]

**Comment by tebjan** [comment]
could you point to a code snipped which loads the midi file? and attach a midi file which fails... also, do you use the nuget (if yes, which version?) or did you compile the sources?

**Comment by tebjan** [comment]
since there is no more info on this i'll close it...

**Comment by stephendrew** [comment]
I also get this, after reading a midi file using File.ReadAllBytes(), passing to a MemoryStream and trying to play it via:

```csharp
var seq = new Sanford.Multimedia.Midi.Sequencer();
seq.Sequence = new Sequence(new MemoryStream(midi));
seq.Start();
```

The error happens creating the sequence.

I am using the latest version in NuGet.

Thanks

**Comment by stephendrew** [comment]
Update: my MIDI file has 70 divisions...

**Comment by tebjan** [comment]
do you know whether this is officially allowed in the midi file specification? if yes, then it is a bug in the library...

**Comment by tebjan** [comment]
it seems that this library expects ppqm to be a multiple of 24:
https://github.com/tebjan/Sanford.Multimedia.Midi/blob/master/Source/Sanford.Multimedia.Midi/Clocks/PpqnClock.cs#L212

this might be a wrong assumption... but i don't know what the midi spec is saying about that. where does the midi file come from? which program has saved it?

**Comment by stephendrew** [comment]
It has been converted from a MUS file from Doom game.

The spec says something like:

> 
> The third word, *division*, specifies the meaning of the delta-times. It has two formats, one for metrical time, and one for time-code-based time:
> 
> bit 15 | bits 14 thru 8 | bits 7 thru 0
> -- | -- | --
> 0 | ticks per quarter-note
> 1 | negative SMPTE format | ticks per frame
> 
> If bit 15 of *division* is zero, the bits 14 thru 0 represent the number of delta time "ticks" which make up a quarter-note. For instance, if division is 96, then a time interval of an eighth-note between two events in the file would be 48.
> 
> If bit 15 of *division* is a one, delta times in a file correspond to subdivisions of a second, in a way consistent with SMPTE and MIDI Time Code. Bits 14 thru 8 contain one of the four values -24, -25, -29, or -30, corresponding to the four standard SMPTE and MIDI Time Code formats (-29 corresponds to 30 drop frame), and represents the number of frames per second. These negative numbers are stored in two's compliment form. The second byte (stored positive) is the resolution within a frame: typical values may be 4 (MIDI Time Code resolution), 8, 10, 80 (bit resolution), or 100. This stream allows exact specifications of time-code-based tracks, but also allows millisecond-based tracks by specifying 25 frames/sec and a resolution of 40 units per frame. If the events in a file are stored with a bit resolution of thirty-frame time code, the division word would be E250 hex.

**Comment by tebjan** [comment]
good find, thanks! i'll see how that can be implemented... if you are able to code, would be a big help if you could have a look. i don't know when i will have the time to look after it.

**Comment by stephendrew** [comment]
I will have a play around with it and see if I can do something.

**Comment by stephendrew** [comment]
I changed `PpqnClock.PpqnMinValue` to be 1 and it seemed to play it fine.  I'm not entirely sure why it was set to 24, and am wary about what it might break :)

**Comment by IainMNorman** [comment]
Interesting. I have a distant recollection of doing something with this before this repo existed. It's very vague. I wonder if I patched it in one uber project I've been using SMM in all these years. I've always just had a copy of the project embedded in my solution, as this nice repo didn't exist back then. I have made a few tweaks to it in years past. I'll look through the commits and see what it is I remember.

-- Edit --

Sorry had a look in my bitbucket repo, it doesn't go back far enough, only to when I switched over from svn, no history. I'll try and remember when I recognise this Ppqn thing. It was something to do with odd time signatures I'm sure.

"I'm not entirely sure why it was set to 24, and am wary about what it might break :)" - That does raise the issue of test coverage :)
